### PR TITLE
[V7] Ring buffer: remove 'old style' version

### PR DESCRIPTION
**Pull request by lucasdemarchi**
continuation of #4593 

Tested on pixhawk.

### CODE REVIEW

File: libraries/DataFlash/DataFlash_File.cpp [hunk 1]
```diff
@@ -1091,36 +1051,34 @@ void DataFlash_File::_io_timer(void)
...
     if ((nbytes + _write_offset) % 512 != 0) {
         uint32_t ofs = (nbytes + _write_offset) % 512;
         if (ofs < nbytes) {
             nbytes -= ofs;
         }
     }
 
-    assert(((uint32_t)_writebuf_head)+nbytes <= _writebuf_size);
-    ssize_t nwritten = ::write(_write_fd, &_writebuf[_writebuf_head], nbytes);
+    ssize_t nwritten = ::write(_write_fd, &head, nbytes);
```
Comment (tridge) [inline libraries/DataFlash/DataFlash_File.cpp]: that doesn't look right to me. We'd write the pointer instead of the data

File: libraries/DataFlash/DataFlash_File.cpp [hunk 2]
```diff
@@ -1091,36 +1051,34 @@ void DataFlash_File::_io_timer(void)
...
         close(_write_fd);
         _write_fd = -1;
-        _initialised = false;
+        _initialized = false;
     } else {
         _write_offset += nwritten;
+        _writebuf.advance(nwritten);
         /*
-          the best strategy for minimising corruption on microSD cards
+          the best strategy for minimizing corruption on microSD cards
```
Comment (tridge) [inline libraries/DataFlash/DataFlash_File.cpp]: I prefer not to have a language war between english/US/australian spellings. Given its been done here then I won't ask it be changed, but please don't switch between different spellings as part of patch sets as it just clutters the patch

**Comment by tridge** [comment]
I've confirmed that running SITL with valgrind crashes as soon as it tries to log any data

File: libraries/DataFlash/DataFlash_File.cpp (hunk 1 above)
Comment (mbelluzzo) [inline libraries/DataFlash/DataFlash_File.cpp]: You are right, '&' survived the sed. Thanks for the notice.

File: libraries/AP_HAL_Linux/RPIOUARTDriver.cpp [hunk 3]
```diff
@@ -75,31 +73,8 @@ void RPIOUARTDriver::begin(uint32_t b, uint16_t rxS, uint16_t txS)
...
-       _writebuf_size = txS;
-       if (_writebuf != NULL) {
-           free(_writebuf);
-       }
-       _writebuf = (uint8_t *)malloc(_writebuf_size);
-       _writebuf_head = 0;
-       _writebuf_tail = 0;
-   }
+    _readbuf.set_size(rxS);
+    _writebuf.set_size(txS);
```
Comment (OXINARF) [inline libraries/AP_HAL_Linux/RPIOUARTDriver.cpp]: Before we only changed size if *txS*/*rxS* were different than 0 and different than current size. This is kept in other places, any reason not to do it here?

File: libraries/AP_HAL_Linux/RPIOUARTDriver.cpp [hunk 4]
```diff
@@ -232,8 +184,7 @@ void RPIOUARTDriver::_timer_tick(void)
     _dma_packet_tx.count_code = 0 | PKT_CODE_READ;
     _dma_packet_tx.page = 0;
     _dma_packet_tx.offset = 0;
-    memset( &_dma_packet_tx.regs[0], 0, PKT_MAX_REGS*sizeof(uint16_t) );
-    _dma_packet_tx.crc = 0;
+    memset(&_dma_packet_tx.regs[0], 0, PKT_MAX_REGS*sizeof(uint16_t) );
```
Comment (OXINARF) [inline libraries/AP_HAL_Linux/RPIOUARTDriver.cpp]: Space before the last parenthesis.

File: libraries/AP_HAL_Linux/SPIUARTDriver.cpp [hunk 5]
```diff
@@ -49,31 +49,8 @@ void SPIUARTDriver::begin(uint32_t b, uint16_t rxS, uint16_t txS)
...
-        _writebuf_size = txS;
-        if (_writebuf != NULL) {
-            free(_writebuf);
-        }
-        _writebuf = (uint8_t *)malloc(_writebuf_size);
-        _writebuf_head = 0;
-        _writebuf_tail = 0;
-    }
+    _readbuf.set_size(rxS);
+    _writebuf.set_size(txS);
```
Comment (OXINARF) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: Same question about `!= 0` and `!= current_size`.

File: libraries/AP_HAL_Linux/SPIUARTDriver.cpp [hunk 6]
```diff
@@ -122,7 +99,7 @@ int SPIUARTDriver::_write_fd(const uint8_t *buf, uint16_t size)
 
     _dev->get_semaphore()->give();
 
-    BUF_ADVANCEHEAD(_writebuf, size);
+    _writebuf.advance(size);
```
Comment (OXINARF) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: Shouldn't this be removed? The advance is now done by the caller right?

File: libraries/AP_HAL_Linux/UARTDriver.cpp [hunk 7]
```diff
@@ -106,54 +105,15 @@ void UARTDriver::_allocate_buffers(uint16_t rxS, uint16_t txS)
...
-        if (_writebuf != NULL) {
-            free(_writebuf);
-        }
-        _writebuf = (uint8_t *)malloc(_writebuf_size);
-        _writebuf_head = 0;
-        _writebuf_tail = 0;
-    }
-
-    if (_writebuf_size != 0 && _readbuf_size != 0) {
+    if (_writebuf.set_size(txS) && _readbuf.set_size(rxS)) {
```
Comment (OXINARF) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: Again, this is different than before:
- as others, it only changed size if *txS*/*rxS* were different than 0 and current size
- it only set *_initialised* if the buffers had a size different than 0

Now this could be setting the buffers to have size 0 and set *_initialised* to true.

File: libraries/AP_HAL_Linux/UARTDriver.cpp [hunk 8]
```diff
@@ -284,45 +244,37 @@ void UARTDriver::set_blocking_writes(bool blocking)
...
   return the number of bytes available to be read
  */
 uint32_t UARTDriver::available()
 {
-    if (!_initialised) {
-        return 0;
-    }
-    uint16_t _tail;
-    return BUF_AVAILABLE(_readbuf);
+    return _readbuf.available();
```
Comment (OXINARF) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: Should the check for *_initialised* really be removed? I mean, it can happen that *_writebuf* was set to the correct size and the *_readbuf* wasn't or vice-versa.

File: libraries/AP_HAL_Linux/UARTDriver.cpp [hunk 9]
```diff
@@ -284,45 +244,37 @@ void UARTDriver::set_blocking_writes(bool blocking)
...
   how many bytes are available in the output buffer?
  */
 uint32_t UARTDriver::txspace()
 {
-    if (!_initialised) {
-        return 0;
-    }
-    uint16_t _head;
-    return BUF_SPACE(_writebuf);
+    return _writebuf.space();
```
Comment (OXINARF) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: Same

File: libraries/AP_HAL_Linux/UARTDriver.cpp [hunk 10]
```diff
@@ -364,45 +313,14 @@ size_t UARTDriver::write(const uint8_t *buffer, size_t size)
...
-    BUF_ADVANCETAIL(_writebuf, n);
-    buffer += n;
-    n = size - n;
-    if (n > 0) {
-        assert(_writebuf_tail+n <= _writebuf_size);
-        memcpy(&_writebuf[_writebuf_tail], buffer, n);
-        BUF_ADVANCETAIL(_writebuf, n);
-    }
-    return size;
+    return _writebuf.write(buffer, size);
```
Comment (OXINARF) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: *ByteBuffer::write* returns an *uint32_t* but this function returns a *size_t*. It won't be a problem now because we also write *size_t*, but this should be improved in the future.

File: libraries/AP_HAL_Linux/UARTDriver.cpp [hunk 11]
```diff
@@ -450,12 +353,11 @@ bool UARTDriver::_write_pending_bytes(void)
...
     // write any pending bytes
-    uint16_t _tail;
-    uint16_t available_bytes = BUF_AVAILABLE(_writebuf);
+    uint32_t available_bytes = _writebuf.available();
     n = available_bytes;
     if (_packetise && n > 0 &&
-        (_writebuf[_writebuf_head] != MAVLINK_STX_MAVLINK1 &&
-         _writebuf[_writebuf_head] != MAVLINK_STX)) {
+        (_writebuf.peek(0) != MAVLINK_STX_MAVLINK1 &&
+         _writebuf.peek(0) != MAVLINK_STX)) {
```
Comment (OXINARF) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: Not a blocker, but this should be changed to the same as below, saving it in a variable.

File: libraries/AP_HAL_Linux/UARTDriver.cpp [hunk 12]
```diff
@@ -475,8 +377,8 @@ bool UARTDriver::_write_pending_bytes(void)
             n = limit;
         }
     }
-    const uint8_t b = _writebuf[_writebuf_head];
-    if (_packetise && n > 0 &&
+    const int16_t b = _writebuf.peek(0);
+    if (_packetise && n > 0 && b > 0 &&
```
Comment (OXINARF) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: Why the `b > 0`? It is already checked that it has to be equal to one of two values.

File: libraries/AP_HAL_Linux/UARTDriver.cpp [hunk 13]
```diff
@@ -508,30 +410,27 @@ bool UARTDriver::_write_pending_bytes(void)
...
-                int ret = _write_fd(&_writebuf[_writebuf_head], n1);
-                if (ret == n1 && n > n1) {
-                    _write_fd(&_writebuf[_writebuf_head], n - n1);
-                }
+            ByteBuffer::IoVec vec[2];
+            const auto n_vec = _writebuf.peekiovec(vec, n);
+            for (int i = 0; i < n_vec; i++) {
+                ret = _write_fd(vec[i].data, (uint16_t)vec[i].len);
+                if (ret > 0)
+                    _writebuf.advance(ret);
```
Comment (OXINARF) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: As for PX4 there should be a check for `ret < vec[i].len`.

File: libraries/AP_HAL_Linux/UARTDriver.cpp [hunk 14]
```diff
@@ -553,22 +450,17 @@ void UARTDriver::_timer_tick(void)
...
+    int ret;
+    ByteBuffer::IoVec vec[2];
+
+    const auto n_vec = _readbuf.reserve(vec, _readbuf.space());
+    for (int i = 0; i < n_vec; i++) {
+        ret = _read_fd(vec[i].data, vec[i].len);
+        if (ret < 0)
+            break;
+        _readbuf.commit((unsigned)ret);
+        if ((unsigned)ret < vec[i].len)
```
Comment (OXINARF) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: Change to the same format as PX4, with the comment?

File: libraries/AP_HAL_PX4/UARTDriver.cpp [hunk 15]
```diff
@@ -93,19 +87,13 @@ void PX4UARTDriver::begin(uint32_t b, uint16_t rxS, uint16_t txS)
...
         }
-		_writebuf_size = txS;
-		if (_writebuf != NULL) {
-			free(_writebuf);
-		}
-		_writebuf = (uint8_t *)malloc(_writebuf_size+16);
-		_writebuf_head = 0;
-		_writebuf_tail = 0;
-	}
+        _writebuf.set_size(txS);
```
Comment (OXINARF) [inline libraries/AP_HAL_PX4/UARTDriver.cpp]: I understand that we can't add 16 here, but do we know why the *malloc* had that 16 added? I just want to make sure this wasn't a workaround for an underlying PX4 bug.

File: libraries/AP_HAL_PX4/UARTDriver.cpp [hunk 16]
```diff
@@ -237,75 +214,69 @@ bool PX4UARTDriver::tx_pending() { return false; }
...
+
+    uint8_t byte;
+    if (!_readbuf.read_byte(&byte)) {
         return -1;
     }
-    c = _readbuf[_readbuf_head];
-    BUF_ADVANCEHEAD(_readbuf, 1);
-	return c;
+
+    return byte;
```
Comment (OXINARF) [inline libraries/AP_HAL_PX4/UARTDriver.cpp]: In Linux there's a cast, let's maintain consistency and add it here too.

File: libraries/AP_HAL_VRBRAIN/UARTDriver.cpp [hunk 17]
```diff
@@ -238,49 +216,45 @@ bool VRBRAINUARTDriver::tx_pending() { return false; }
...
+
+    uint8_t byte;
+    if (!_readbuf.read_byte(&byte))
         return -1;
-    }
-    c = _readbuf[_readbuf_head];
-    BUF_ADVANCEHEAD(_readbuf, 1);
-	return c;
+
+    return byte;
```
Comment (OXINARF) [inline libraries/AP_HAL_VRBRAIN/UARTDriver.cpp]: Cast to maintain consistency

File: libraries/AP_HAL_VRBRAIN/UARTDriver.h [hunk 18]
```diff
@@ -48,18 +50,9 @@ class VRBRAIN::VRBRAINUARTDriver : public AP_HAL::UARTDriver {
...
-    // data is put
-    volatile uint16_t _readbuf_head;
-    volatile uint16_t _readbuf_tail;
-
-    uint8_t *_writebuf;
-    uint16_t _writebuf_size;
-    volatile uint16_t _writebuf_head;
-    volatile uint16_t _writebuf_tail;
+    ByteBuffer _readbuf;
+    ByteBuffer _writebuf;
```
Comment (OXINARF) [inline libraries/AP_HAL_VRBRAIN/UARTDriver.h]: This is the only HAL that these aren't initialized, any reason why?

File: libraries/AP_HAL_Linux/RPIOUARTDriver.cpp (hunk 4 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/RPIOUARTDriver.cpp]: fixing this

File: libraries/AP_HAL_Linux/SPIUARTDriver.cpp (hunk 5 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: it's never 0, see lines above.  And it doesn't change if the size is the same due to the implementation of set_size().

File: libraries/AP_HAL_Linux/SPIUARTDriver.cpp (hunk 6 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: yes, you are right... I think I focused on carefully reviewing the PX4 version and forgot the others.

File: libraries/AP_HAL_Linux/UARTDriver.cpp (hunk 13 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: indeed, I'll apply the same fixes I did for PX4 to the others

File: libraries/AP_HAL_Linux/UARTDriver.cpp (hunk 14 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: :+1:

File: libraries/AP_HAL_PX4/UARTDriver.cpp (hunk 15 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_PX4/UARTDriver.cpp]: @tridge can explain since he's the original author. But the only reason I can imagine is to serve as a "guard size" to protect against bugs in the ringbuffer implementation.  So if you write past your size, you will corrupt the packet, but at least you won't corrupt other data in memory (or crash in some cases).  I don't see a need to keep the previous behavior

File: libraries/AP_HAL_Linux/RPIOUARTDriver.cpp (hunk 3 above)
Comment (OXINARF) [inline libraries/AP_HAL_Linux/RPIOUARTDriver.cpp]: Ok, I missed that it was enforcing a minimum of 1024.
I didn't understand what you are saying about other places.

File: libraries/AP_HAL_Linux/UARTDriver.cpp (hunk 12 above)
Comment (OXINARF) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: No, it wasn't.

File: libraries/AP_HAL/utility/RingBuffer.cpp [hunk 19]
```diff
@@ -47,8 +53,22 @@ void ByteBuffer::clear(void)
...
 uint32_t ByteBuffer::space(void) const
 {
+    if (size == 0) {
+        return 0;
+    }
+
+    /* use a copy on stack to avoid race conditions of @head being updated by
+     * the reader thread */
     uint32_t _head = head;
-    return size ? (_head > tail ? 0 : size) + _head - tail - 1 : 0;
```
Comment (dgrat) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: I actually like this line more than the if clauses below ..

File: libraries/AP_HAL/utility/RingBuffer.cpp [hunk 20]
```diff
@@ -196,6 +216,16 @@ uint32_t ByteBuffer::read(uint8_t *data, uint32_t len)
     return ret;
 }
 
+bool ByteBuffer::read_byte(uint8_t *data)
```
Comment (dgrat) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: There is no check for the pointer

File: libraries/AP_HAL/utility/RingBuffer.cpp [hunk 21]
```diff
@@ -196,6 +216,16 @@ uint32_t ByteBuffer::read(uint8_t *data, uint32_t len)
...
+{
+    int16_t ret = peek(0);
+    if (ret < 0) {
+        return false;
+    }
+    *data = ret;
```
Comment (dgrat) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: Which may led to crash here :D

File: libraries/AP_HAL_Linux/RPIOUARTDriver.cpp [hunk 22]
```diff
@@ -186,42 +160,20 @@ void RPIOUARTDriver::_timer_tick(void)
...
     }
 
-    struct IOPacket _dma_packet_tx, _dma_packet_rx;
+    struct IOPacket _dma_packet_tx = {0}, _dma_packet_rx = {0};
 
     /* get write_buf bytes */
-    uint16_t _tail;
-    uint16_t n = BUF_AVAILABLE(_writebuf);
+    uint32_t n = MIN(_writebuf.available(), PKT_MAX_REGS * 2);
+    n = MIN(n, _baudrate / 10U / (1000000U / RPIOUART_POLL_TIME_INTERVAL));
```
Comment (dgrat) [inline libraries/AP_HAL_Linux/RPIOUARTDriver.cpp]: This reads cryptic

File: libraries/AP_HAL_Linux/SPIUARTDriver.cpp [hunk 23]
```diff
@@ -131,49 +108,15 @@ int SPIUARTDriver::_write_fd(const uint8_t *buf, uint16_t size)
...
+    _readbuf.write(_buffer, size);
 
     return ret;
 }
 
-static uint8_t ff_stub[300] = {0xff};
-
 int SPIUARTDriver::_read_fd(uint8_t *buf, uint16_t n)
 {
+    static uint8_t ff_stub[100] = {0xff};
```
Comment (dgrat) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: Actually, the proper way would be to make a member variable out of it. 
Mixing C- and C++ code looks not nice.

File: libraries/AP_HAL_Linux/UARTDriver.cpp [hunk 24]
```diff
@@ -284,45 +244,37 @@ void UARTDriver::set_blocking_writes(bool blocking)
...
+
+    uint8_t byte;
+    if (!_readbuf.read_byte(&byte)) {
         return -1;
     }
-    c = _readbuf[_readbuf_head];
-    BUF_ADVANCEHEAD(_readbuf, 1);
-    return c;
+
+    return (int16_t) byte;
```
Comment (dgrat) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: it would get converted anyway to this type.

File: libraries/AP_HAL_Linux/UARTDriver.cpp [hunk 25]
```diff
@@ -364,45 +313,14 @@ size_t UARTDriver::write(const uint8_t *buffer, size_t size)
...
-        BUF_ADVANCETAIL(_writebuf, n);
-    }
-    return size;
+    return _writebuf.write(buffer, size);
 }
 
 /*
   try writing n bytes, handling an unresponsive port
  */
 int UARTDriver::_write_fd(const uint8_t *buf, uint16_t n)
```
Comment (dgrat) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: again no pointer check.

File: libraries/AP_HAL_Linux/UARTDriver.cpp [hunk 26]
```diff
@@ -414,30 +332,15 @@ int UARTDriver::_write_fd(const uint8_t *buf, uint16_t n)
...
-    }
-
-    return ret;
+    return _device->write(buf, n);
 }
 
 /*
   try reading n bytes, handling an unresponsive port
  */
 int UARTDriver::_read_fd(uint8_t *buf, uint16_t n)
```
Comment (dgrat) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: no pointer checks

**Comment by dgrat** [comment]
I was just checking a fraction and I like this commit.
However, there are zero pointer/buffer checks.
Conseqeuntly, the code is unsafe. 
It would be smart to use asserts leading to compiler errors in case the function are not correctly used.

File: libraries/AP_HAL_PX4/UARTDriver.cpp [hunk 27]
```diff
@@ -500,47 +420,46 @@ void PX4UARTDriver::_timer_tick(void)
     _in_timer = true;
 
     // write any pending bytes
-    uint16_t _tail;
-    n = BUF_AVAILABLE(_writebuf);
+    n = _writebuf.available();
```
Comment (dgrat) [inline libraries/AP_HAL_PX4/UARTDriver.cpp]: we don't really need this variable 'n', because it is just used 1,2 times.

File: libraries/AP_HAL_VRBRAIN/UARTDriver.cpp (hunk 17 above)
Comment (dgrat) [inline libraries/AP_HAL_VRBRAIN/UARTDriver.cpp]: Standard return type might be -1. This would save some lines of code.

File: libraries/AP_HAL_VRBRAIN/UARTDriver.cpp [hunk 28]
```diff
@@ -295,17 +269,14 @@ size_t VRBRAINUARTDriver::write(uint8_t c)
         try_initialise();
         return 0;
     }
-    uint16_t _head;
 
-    while (BUF_SPACE(_writebuf) == 0) {
+    while (_writebuf.space() == 0) {
```
Comment (dgrat) [inline libraries/AP_HAL_VRBRAIN/UARTDriver.cpp]: the if could be in while, making the function shorter

File: libraries/AP_HAL_Linux/SPIUARTDriver.cpp (hunk 23 above)
Comment (magicrub) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: static variables are not allowed, please make a member function. if for some reason this class gets reused in the future, a static variable would cause incorrect behavior. Using a member variable takes up the same amount of space (in single class case) and is safer.

Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: Please, read the diff again. It was already static, it just has a reduced scope now since it's inside the function.

Btw, this is a scratch buffer. since it needs to pass a value != 0 as transmit buffer (this is only to maintain the previous behavior since I don't actually know if the real hardware is ok with 0s instead of 0xFFs).

Comment (magicrub) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: still needs to change, right? Just because it was wrong before doesn't mean it's OK to keep it.

Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: It's actually not wrong... Static const var here serves the same purpose as a static const var inside the class. I'd rather remove this change, leave as it was an let the fix to be in another PR.

Comment (magicrub) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: I would agree that the change in this PR doesn't help, only makes it worse, and reverting to old behavior and fixing in new PR is OK.

File: libraries/AP_HAL_Linux/RPIOUARTDriver.cpp (hunk 3 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/RPIOUARTDriver.cpp]: you said that "This is kept in other places, any reason not to do it here?" and I mentioned that in these other places there are other reasons to do the check, like setting a `initialized` variable to false.

File: libraries/AP_HAL_Linux/RPIOUARTDriver.cpp (hunk 4 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/RPIOUARTDriver.cpp]: I actually removed this commit.

File: libraries/AP_HAL_Linux/UARTDriver.cpp (hunk 7 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: same as in other places... we don't allow size == 0 in this method. And `set_size` takes care of it being the same as the previous size.

File: libraries/AP_HAL_Linux/UARTDriver.cpp (hunk 25 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: this is internal method

File: libraries/AP_HAL_Linux/UARTDriver.cpp (hunk 26 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: internal method

File: libraries/AP_HAL_Linux/UARTDriver.cpp (hunk 11 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: fixing...

File: libraries/AP_HAL_Linux/UARTDriver.cpp (hunk 24 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: Ok, removing the cast

File: libraries/AP_HAL_Linux/UARTDriver.cpp (hunk 10 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: I'll leave this for later. On every arch we support size_t is >= uint32_t so no big deal

File: libraries/AP_HAL_Linux/UARTDriver.cpp (hunk 12 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/UARTDriver.cpp]: Fixed

File: libraries/AP_HAL_VRBRAIN/UARTDriver.h (hunk 18 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_VRBRAIN/UARTDriver.h]: None of them need the initialization actuallu

File: libraries/AP_HAL_PX4/UARTDriver.cpp [hunk 29]
```diff
@@ -72,19 +71,14 @@ void PX4UARTDriver::begin(uint32_t b, uint16_t rxS, uint16_t txS)
       thrashing of the heap once we are up. The ttyACM0 driver may not
       connect for some time after boot
      */
-	if (rxS != 0 && rxS != _readbuf_size) {
+    if (rxS != 0 && rxS != _readbuf.get_size()) {
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_PX4/UARTDriver.cpp]: Need to take care of the _initialized flag that's a bit different than other places

File: libraries/AP_HAL/utility/RingBuffer.cpp (hunk 19 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: I removed this commit, but these 2 ternary operators are unreadable

File: libraries/AP_HAL/utility/RingBuffer.cpp (hunk 20 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: Added a check here since this is public method

File: libraries/AP_HAL/utility/RingBuffer.cpp (hunk 21 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: Ok

File: libraries/AP_HAL_Linux/SPIUARTDriver.cpp (hunk 23 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: Removed this commit

File: libraries/AP_HAL_VRBRAIN/UARTDriver.cpp (hunk 28 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_VRBRAIN/UARTDriver.cpp]: I couldn't understand this. We return 0 in the  `if` branch

File: libraries/DataFlash/DataFlash_File.cpp (hunk 1 above)
Comment (lucasdemarchi) [inline libraries/DataFlash/DataFlash_File.cpp]: Fixed

File: libraries/DataFlash/DataFlash_File.cpp [hunk 30]
```diff
@@ -1091,36 +1051,34 @@ void DataFlash_File::_io_timer(void)
...
 
-    assert(((uint32_t)_writebuf_head)+nbytes <= _writebuf_size);
-    ssize_t nwritten = ::write(_write_fd, &_writebuf[_writebuf_head], nbytes);
+    ssize_t nwritten = ::write(_write_fd, &head, nbytes);
     if (nwritten <= 0) {
         hal.util->perf_count(_perf_errors);
         close(_write_fd);
         _write_fd = -1;
-        _initialised = false;
+        _initialized = false;
```
Comment (lucasdemarchi) [inline libraries/DataFlash/DataFlash_File.cpp]: I removed these spell changes

File: libraries/DataFlash/DataFlash_File.cpp (hunk 2 above)
Comment (lucasdemarchi) [inline libraries/DataFlash/DataFlash_File.cpp]: Fixed

**Comment by lucasdemarchi** [comment]
I'll update this pr tomorrow with the fixes mentioned

File: libraries/AP_HAL_VRBRAIN/UARTDriver.cpp (hunk 28 above)
Comment (dgrat) [inline libraries/AP_HAL_VRBRAIN/UARTDriver.cpp]: The if could be moved out of the loop I guess.

File: libraries/AP_HAL/utility/RingBuffer.cpp (hunk 19 above)
Comment (OXINARF) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: @lucasdemarchi Yes, please put the commit back on!

Comment (lucasdemarchi) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: @OXINARF I'll send a separate PR for the coding style changes.... they are just removing the focus of this PR and I want to get this PR applied soon

**Comment by tridge** [comment]
as far I can see this is ready for master. I'm doing some final tests now before pushing

**Comment by tridge** [comment]
tests all passed, pushed to master
many thanks for your work on this! It is great to see the old macros gone

**Comment by mbelluzzo** [comment]
\o/

**Comment by magicrub** [comment]
OMG OMG OMG It's merged!!!!
